File: julia/src/TestGenerator.jl
"""
    TestGenerator

Adaptive test generation based on action types.

Instead of only running existing tests, we generate targeted unit tests
specific to each action. This provides richer training signal for the JEPA
world model by testing the actual consequences of each edit.

Inspired by Agent2World (2025): "A Testing Team conducts adaptive unit testing
and simulation-based validation... dynamically synthesizes test cases based on
the specific errors exhibited by each world model."
"""
module TestGenerator

using Test
using Random

export generate_tests_for_action, TestCase, TestSuite, run_generated_tests
export ActionTestStrategy, MethodAdditionStrategy, MethodModificationStrategy
export FieldAdditionStrategy, ImportStrategy, RenameStrategy

# -----------------------------------------------------------------------------
# Test Case Representation
# -----------------------------------------------------------------------------

"""
A single generated test case.
"""
struct TestCase
    name::String
    code::String
    expected_behavior::Symbol  # :should_pass, :should_error, :should_return
    expected_value::Any        # For :should_return cases
    timeout_ms::Int
end

"""
A suite of test cases for a specific action.
"""
struct TestSuite
    action_type::Symbol
    target::String
    tests::Vector{TestCase}
    setup_code::String
    teardown_code::String
end

# -----------------------------------------------------------------------------
# Test Generation Strategies (per action type)
# -----------------------------------------------------------------------------

abstract type ActionTestStrategy end

"""
Strategy for testing ADD_METHOD actions.
"""
struct MethodAdditionStrategy <: ActionTestStrategy
    method_name::String
    module_name::String
    arg_types::Vector{Type}
    return_type::Union{Type, Nothing}
end

"""
Strategy for testing MODIFY_METHOD actions.
"""
struct MethodModificationStrategy <: ActionTestStrategy
    method_name::String
    module_name::String
    original_signature::String
    new_signature::String
end

"""
Strategy for testing ADD_FIELD actions.
"""
struct FieldAdditionStrategy <: ActionTestStrategy
    struct_name::String
    field_name::String
    field_type::Type
end

"""
Strategy for testing ADD_IMPORT / REMOVE_IMPORT actions.
"""
struct ImportStrategy <: ActionTestStrategy
    module_name::String
    imported_symbols::Vector{Symbol}
    is_addition::Bool
end

"""
Strategy for testing RENAME_SYMBOL actions.
"""
struct RenameStrategy <: ActionTestStrategy
    old_name::String
    new_name::String
    symbol_type::Symbol  # :function, :type, :module, :variable
end

# -----------------------------------------------------------------------------
# Test Generation Functions
# -----------------------------------------------------------------------------

"""
    generate_tests_for_action(action_type, target, context) -> TestSuite

Generate targeted tests based on the action type and target.

# Arguments
- `action_type::Symbol`: One of :ADD_METHOD, :MODIFY_METHOD, :REMOVE_METHOD, etc.
- `target::String`: The target of the action (method name, field name, etc.)
- `context::Dict`: Additional context (module, types, existing tests, etc.)

# Returns
A TestSuite with tests specific to this action.
"""
function generate_tests_for_action(
    action_type::Symbol,
    target::String,
    context::Dict
)::TestSuite
    
    strategy = create_strategy(action_type, target, context)
    tests = generate_tests(strategy, context)
    
    TestSuite(
        action_type,
        target,
        tests,
        get(context, :setup_code, ""),
        get(context, :teardown_code, "")
    )
end

"""
Create the appropriate strategy for the action type.
"""
function create_strategy(action_type::Symbol, target::String, context::Dict)
    if action_type == :ADD_METHOD
        return MethodAdditionStrategy(
            target,
            get(context, :module_name, "Main"),
            get(context, :arg_types, Type[]),
            get(context, :return_type, nothing)
        )
    elseif action_type == :MODIFY_METHOD
        return MethodModificationStrategy(
            target,
            get(context, :module_name, "Main"),
            get(context, :original_signature, ""),
            get(context, :new_signature, "")
        )
    elseif action_type == :ADD_FIELD
        return FieldAdditionStrategy(
            get(context, :struct_name, ""),
            target,
            get(context, :field_type, Any)
        )
    elseif action_type in (:ADD_IMPORT, :REMOVE_IMPORT)
        return ImportStrategy(
            get(context, :module_name, ""),
            get(context, :imported_symbols, Symbol[]),
            action_type == :ADD_IMPORT
        )
    elseif action_type == :RENAME_SYMBOL
        return RenameStrategy(
            get(context, :old_name, ""),
            target,
            get(context, :symbol_type, :function)
        )
    else
        # Default: return empty tests for unhandled action types
        return nothing
    end
end

# -----------------------------------------------------------------------------
# Strategy-Specific Test Generation
# -----------------------------------------------------------------------------

"""
Generate tests for method addition.
"""
function generate_tests(strategy::MethodAdditionStrategy, context::Dict)::Vector{TestCase}
    tests = TestCase[]
    method_name = strategy.method_name
    module_name = strategy.module_name
    arg_types = strategy.arg_types
    
    # Test 1: Method exists and is callable
    push!(tests, TestCase(
        "$(method_name)_exists",
        """
        @test isdefined($module_name, :$method_name)
        @test hasmethod($module_name.$method_name, Tuple{$(join(string.(arg_types), ", "))})
        """,
        :should_pass,
        nothing,
        5000
    ))
    
    # Test 2: Method can be called with correct argument types
    if !isempty(arg_types)
        sample_args = generate_sample_args(arg_types)
        push!(tests, TestCase(
            "$(method_name)_callable",
            """
            args = $sample_args
            result = $module_name.$method_name(args...)
            @test result !== nothing || true  # Just test it doesn't error
            """,
            :should_pass,
            nothing,
            10000
        ))
    end
    
    # Test 3: Method handles edge cases (if numeric types)
    if any(t -> t <: Number, arg_types)
        push!(tests, TestCase(
            "$(method_name)_edge_cases",
            """
            # Test with zeros
            zero_args = $(generate_zero_args(arg_types))
            @test try $module_name.$method_name(zero_args...); true catch; true end
            """,
            :should_pass,
            nothing,
            5000
        ))
    end
    
    # Test 4: Return type check (if specified)
    if strategy.return_type !== nothing
        sample_args = generate_sample_args(arg_types)
        push!(tests, TestCase(
            "$(method_name)_return_type",
            """
            args = $sample_args
            result = $module_name.$method_name(args...)
            @test result isa $(strategy.return_type)
            """,
            :should_pass,
            nothing,
            5000
        ))
    end
    
    # Test 5: Method doesn't break existing functionality
    # (Integration test with existing methods)
    if haskey(context, :related_methods)
        for related in context[:related_methods]
            push!(tests, TestCase(
                "$(method_name)_compatible_with_$(related)",
                """
                # Ensure adding $method_name doesn't break $related
                @test hasmethod($module_name.$related, Tuple{})
                """,
                :should_pass,
                nothing,
                5000
            ))
        end
    end
    
    tests
end

"""
Generate tests for method modification.
"""
function generate_tests(strategy::MethodModificationStrategy, context::Dict)::Vector{TestCase}
    tests = TestCase[]
    method_name = strategy.method_name
    module_name = strategy.module_name
    
    # Test 1: Method still exists after modification
    push!(tests, TestCase(
        "$(method_name)_still_exists",
        """
        @test isdefined($module_name, :$method_name)
        """,
        :should_pass,
        nothing,
        5000
    ))
    
    # Test 2: New signature is valid
    if !isempty(strategy.new_signature)
        push!(tests, TestCase(
            "$(method_name)_new_signature_valid",
            """
            # Check method can be called with new signature
            methods_list = methods($module_name.$method_name)
            @test length(methods_list) > 0
            """,
            :should_pass,
            nothing,
            5000
        ))
    end
    
    # Test 3: Regression test - existing callers still work
    if haskey(context, :existing_call_sites)
        for (i, call_site) in enumerate(context[:existing_call_sites])
            push!(tests, TestCase(
                "$(method_name)_regression_$i",
                """
                # Existing call site should still work
                @test try $call_site; true catch e; false end
                """,
                :should_pass,
                nothing,
                10000
            ))
        end
    end
    
    # Test 4: Check for type stability (important for Julia)
    push!(tests, TestCase(
        "$(method_name)_type_stable",
        """
        # Type stability check placeholder
        # In real implementation, use @code_warntype
        @test true
        """,
        :should_pass,
        nothing,
        5000
    ))
    
    tests
end

"""
Generate tests for field addition.
"""
function generate_tests(strategy::FieldAdditionStrategy, context::Dict)::Vector{TestCase}
    tests = TestCase[]
    struct_name = strategy.struct_name
    field_name = strategy.field_name
    field_type = strategy.field_type
    
    # Test 1: Field exists on the struct
    push!(tests, TestCase(
        "$(struct_name)_has_$(field_name)",
        """
        @test hasfield($struct_name, :$field_name)
        @test fieldtype($struct_name, :$field_name) == $field_type
        """,
        :should_pass,
        nothing,
        5000
    ))
    
    # Test 2: Struct can still be constructed
    if haskey(context, :constructor_args)
        push!(tests, TestCase(
            "$(struct_name)_constructible",
            """
            args = $(context[:constructor_args])
            obj = $struct_name(args...)
            @test obj.$field_name !== nothing || true
            """,
            :should_pass,
            nothing,
            5000
        ))
    end
    
    # Test 3: Field is accessible
    push!(tests, TestCase(
        "$(struct_name)_$(field_name)_accessible",
        """
        # Field should be accessible via getfield
        @test :$field_name in fieldnames($struct_name)
        """,
        :should_pass,
        nothing,
        5000
    ))
    
    tests
end

"""
Generate tests for import changes.
"""
function generate_tests(strategy::ImportStrategy, context::Dict)::Vector{TestCase}
    tests = TestCase[]
    module_name = strategy.module_name
    symbols = strategy.imported_symbols
    
    if strategy.is_addition
        # Test: Imported symbols are now available
        for sym in symbols
            push!(tests, TestCase(
                "import_$(sym)_available",
                """
                @test isdefined($module_name, :$sym)
                """,
                :should_pass,
                nothing,
                5000
            ))
        end
    else
        # Test: Removed imports don't break dependent code
        if haskey(context, :dependent_code)
            push!(tests, TestCase(
                "removal_doesnt_break_deps",
                """
                # Check that removing import doesn't break anything
                # This would fail if dependent code used the import
                @test true  # Placeholder
                """,
                :should_pass,
                nothing,
                5000
            ))
        end
    end
    
    tests
end

"""
Generate tests for symbol renaming.
"""
function generate_tests(strategy::RenameStrategy, context::Dict)::Vector{TestCase}
    tests = TestCase[]
    old_name = strategy.old_name
    new_name = strategy.new_name
    sym_type = strategy.symbol_type
    
    # Test 1: New name exists
    push!(tests, TestCase(
        "$(new_name)_exists",
        """
        @test isdefined(Main, :$new_name)
        """,
        :should_pass,
        nothing,
        5000
    ))
    
    # Test 2: Old name is gone (unless aliased)
    if !get(context, :keep_alias, false)
        push!(tests, TestCase(
            "$(old_name)_removed",
            """
            @test !isdefined(Main, :$old_name)
            """,
            :should_pass,
            nothing,
            5000
        ))
    end
    
    # Test 3: All references updated
    if haskey(context, :reference_sites)
        for (i, site) in enumerate(context[:reference_sites])
            push!(tests, TestCase(
                "reference_$i_updated",
                """
                # Reference site should use new name
                @test contains("$site", "$new_name")
                """,
                :should_pass,
                nothing,
                5000
            ))
        end
    end
    
    tests
end

"""
Fallback for unhandled strategies.
"""
function generate_tests(strategy::Nothing, context::Dict)::Vector{TestCase}
    TestCase[]
end

# -----------------------------------------------------------------------------
# Test Execution
# -----------------------------------------------------------------------------

"""
    run_generated_tests(suite::TestSuite) -> (passed, failed, results)

Execute a test suite and return results.
"""
function run_generated_tests(suite::TestSuite)
    passed = 0
    failed = 0
    results = Dict{String, Any}()
    
    # Run setup
    if !isempty(suite.setup_code)
        try
            eval(Meta.parse(suite.setup_code))
        catch e
            @warn "Setup failed" exception=e
        end
    end
    
    # Run each test
    for test in suite.tests
        result = run_single_test(test)
        results[test.name] = result
        
        if result[:passed]
            passed += 1
        else
            failed += 1
        end
    end
    
    # Run teardown
    if !isempty(suite.teardown_code)
        try
            eval(Meta.parse(suite.teardown_code))
        catch e
            @warn "Teardown failed" exception=e
        end
    end
    
    (passed, failed, results)
end

"""
Run a single test case with timeout.
"""
function run_single_test(test::TestCase)
    start_time = time()
    
    try
        # Parse and evaluate the test code
        expr = Meta.parse("begin\n$(test.code)\nend")
        
        # Run with timeout (simplified - real impl would use async)
        result = eval(expr)
        
        elapsed_ms = (time() - start_time) * 1000
        
        Dict(
            :passed => true,
            :elapsed_ms => elapsed_ms,
            :error => nothing,
            :result => result
        )
    catch e
        elapsed_ms = (time() - start_time) * 1000
        
        # Check if error was expected
        if test.expected_behavior == :should_error
            Dict(
                :passed => true,
                :elapsed_ms => elapsed_ms,
                :error => e,
                :result => nothing
            )
        else
            Dict(
                :passed => false,
                :elapsed_ms => elapsed_ms,
                :error => e,
                :result => nothing
            )
        end
    end
end

# -----------------------------------------------------------------------------
# Helper Functions for Sample Generation
# -----------------------------------------------------------------------------

"""
Generate sample arguments for given types.
"""
function generate_sample_args(types::Vector{Type})
    args = Any[]
    for T in types
        push!(args, generate_sample_value(T))
    end
    args
end

"""
Generate a sample value for a type.
"""
function generate_sample_value(T::Type)
    if T <: Integer
        return 1
    elseif T <: AbstractFloat
        return 1.0
    elseif T <: AbstractString
        return "test"
    elseif T <: AbstractVector
        return []
    elseif T <: AbstractDict
        return Dict()
    elseif T <: Bool
        return true
    elseif T <: Symbol
        return :test
    else
        return nothing
    end
end

"""
Generate zero/empty values for types (edge case testing).
"""
function generate_zero_args(types::Vector{Type})
    args = Any[]
    for T in types
        if T <: Number
            push!(args, zero(T))
        elseif T <: AbstractString
            push!(args, "")
        elseif T <: AbstractVector
            push!(args, T())
        else
            push!(args, nothing)
        end
    end
    args
end

# -----------------------------------------------------------------------------
# Property-Based Test Generation
# -----------------------------------------------------------------------------

"""
    generate_property_tests(method_name, properties) -> Vector{TestCase}

Generate property-based tests that should hold for any input.
"""
function generate_property_tests(
    method_name::String,
    properties::Vector{Symbol},
    context::Dict
)::Vector{TestCase}
    tests = TestCase[]
    
    for prop in properties
        if prop == :idempotent
            push!(tests, TestCase(
                "$(method_name)_idempotent",
                """
                x = $(get(context, :sample_input, "1"))
                @test $method_name($method_name(x)) == $method_name(x)
                """,
                :should_pass,
                nothing,
                10000
            ))
        elseif prop == :commutative
            push!(tests, TestCase(
                "$(method_name)_commutative",
                """
                x, y = $(get(context, :sample_inputs, "(1, 2)"))
                @test $method_name(x, y) == $method_name(y, x)
                """,
                :should_pass,
                nothing,
                10000
            ))
        elseif prop == :associative
            push!(tests, TestCase(
                "$(method_name)_associative",
                """
                x, y, z = $(get(context, :sample_inputs, "(1, 2, 3)"))
                @test $method_name($method_name(x, y), z) == $method_name(x, $method_name(y, z))
                """,
                :should_pass,
                nothing,
                10000
            ))
        elseif prop == :invertible
            push!(tests, TestCase(
                "$(method_name)_invertible",
                """
                x = $(get(context, :sample_input, "1"))
                inv_method = $(get(context, :inverse_method, "inv_$method_name"))
                @test inv_method($method_name(x)) ≈ x
                """,
                :should_pass,
                nothing,
                10000
            ))
        end
    end
    
    tests
end

end # module

File: agent/test_generator.py
"""
Adaptive test generation for JEPA training.

This module generates targeted tests based on action types, providing
richer training signal than running existing tests alone.

Inspired by Agent2World (2025): Testing Team "dynamically synthesizes
test cases based on the specific errors exhibited by each world model,
enabling precision-guided debugging rather than generic checks."
"""

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Optional

# ---------------------------------------------------------------------------
# Test Case Representation
# ---------------------------------------------------------------------------


class ExpectedBehavior(Enum):
    """Expected behavior of a test."""
    SHOULD_PASS = "should_pass"
    SHOULD_ERROR = "should_error"
    SHOULD_RETURN = "should_return"


@dataclass
class TestCase:
    """A single generated test case."""
    name: str
    code: str
    expected_behavior: ExpectedBehavior = ExpectedBehavior.SHOULD_PASS
    expected_value: Any = None
    timeout_ms: int = 5000


@dataclass
class TestResult:
    """Result of running a test case."""
    name: str
    passed: bool
    elapsed_ms: float
    error: Optional[str] = None
    actual_value: Any = None


@dataclass
class TestSuite:
    """A suite of test cases for a specific action."""
    action_type: str
    target: str
    tests: list[TestCase] = field(default_factory=list)
    setup_code: str = ""
    teardown_code: str = ""


# ---------------------------------------------------------------------------
# Test Generation Strategies
# ---------------------------------------------------------------------------


class TestGenerationStrategy:
    """Base class for action-specific test generation."""
    
    def generate(self, context: dict) -> list[TestCase]:
        raise NotImplementedError


class AddMethodStrategy(TestGenerationStrategy):
    """Generate tests for ADD_METHOD actions."""
    
    def __init__(
        self,
        method_name: str,
        module_name: str,
        arg_types: list[str],
        return_type: Optional[str] = None,
    ):
        self.method_name = method_name
        self.module_name = module_name
        self.arg_types = arg_types
        self.return_type = return_type
    
    def generate(self, context: dict) -> list[TestCase]:
        tests = []
        
        # Test 1: Method exists
        tests.append(TestCase(
            name=f"{self.method_name}_exists",
            code=f"""
@test isdefined({self.module_name}, :{self.method_name})
""".strip(),
        ))
        
        # Test 2: Method is callable with correct types
        if self.arg_types:
            type_tuple = ", ".join(self.arg_types)
            tests.append(TestCase(
                name=f"{self.method_name}_has_method",
                code=f"""
@test hasmethod({self.module_name}.{self.method_name}, Tuple{{{type_tuple}}})
""".strip(),
            ))
        
        # Test 3: Basic invocation
        sample_args = self._generate_sample_args()
        if sample_args:
            tests.append(TestCase(
                name=f"{self.method_name}_callable",
                code=f"""
result = {self.module_name}.{self.method_name}({sample_args})
@test result !== nothing || true
""".strip(),
                timeout_ms=10000,
            ))
        
        # Test 4: Return type (if specified)
        if self.return_type and sample_args:
            tests.append(TestCase(
                name=f"{self.method_name}_return_type",
                code=f"""
result = {self.module_name}.{self.method_name}({sample_args})
@test result isa {self.return_type}
""".strip(),
            ))
        
        # Test 5: Edge cases for numeric types
        if any(t in ["Int", "Float64", "Number"] for t in self.arg_types):
            zero_args = self._generate_zero_args()
            tests.append(TestCase(
                name=f"{self.method_name}_handles_zeros",
                code=f"""
@test try {self.module_name}.{self.method_name}({zero_args}); true catch; true end
""".strip(),
            ))
        
        return tests
    
    def _generate_sample_args(self) -> str:
        """Generate sample arguments based on types."""
        samples = []
        for t in self.arg_types:
            if t in ["Int", "Int64", "Integer"]:
                samples.append("1")
            elif t in ["Float64", "Float32", "AbstractFloat"]:
                samples.append("1.0")
            elif t in ["String", "AbstractString"]:
                samples.append('"test"')
            elif t.startswith("Vector"):
                samples.append("[]")
            elif t == "Bool":
                samples.append("true")
            elif t == "Symbol":
                samples.append(":test")
            else:
                samples.append("nothing")
        return ", ".join(samples)
    
    def _generate_zero_args(self) -> str:
        """Generate zero/empty arguments for edge case testing."""
        zeros = []
        for t in self.arg_types:
            if t in ["Int", "Int64", "Integer"]:
                zeros.append("0")
            elif t in ["Float64", "Float32", "AbstractFloat"]:
                zeros.append("0.0")
            elif t in ["String", "AbstractString"]:
                zeros.append('""')
            else:
                zeros.append("nothing")
        return ", ".join(zeros)


class ModifyMethodStrategy(TestGenerationStrategy):
    """Generate tests for MODIFY_METHOD actions."""
    
    def __init__(
        self,
        method_name: str,
        module_name: str,
        original_signature: str = "",
        new_signature: str = "",
    ):
        self.method_name = method_name
        self.module_name = module_name
        self.original_signature = original_signature
        self.new_signature = new_signature
    
    def generate(self, context: dict) -> list[TestCase]:
        tests = []
        
        # Test 1: Method still exists
        tests.append(TestCase(
            name=f"{self.method_name}_still_exists",
            code=f"""
@test isdefined({self.module_name}, :{self.method_name})
""".strip(),
        ))
        
        # Test 2: Has at least one method
        tests.append(TestCase(
            name=f"{self.method_name}_has_methods",
            code=f"""
@test length(methods({self.module_name}.{self.method_name})) > 0
""".strip(),
        ))
        
        # Test 3: Regression tests from existing call sites
        existing_calls = context.get("existing_call_sites", [])
        for i, call in enumerate(existing_calls):
            tests.append(TestCase(
                name=f"{self.method_name}_regression_{i}",
                code=f"""
@test try {call}; true catch; false end
""".strip(),
                timeout_ms=10000,
            ))
        
        return tests


class RemoveMethodStrategy(TestGenerationStrategy):
    """Generate tests for REMOVE_METHOD actions."""
    
    def __init__(self, method_name: str, module_name: str):
        self.method_name = method_name
        self.module_name = module_name
    
    def generate(self, context: dict) -> list[TestCase]:
        tests = []
        
        # Test 1: Method no longer exists (or has fewer methods)
        tests.append(TestCase(
            name=f"{self.method_name}_removed",
            code=f"""
# Either undefined or has fewer methods than before
methods_count = isdefined({self.module_name}, :{self.method_name}) ? 
    length(methods({self.module_name}.{self.method_name})) : 0
@test methods_count < {context.get('original_method_count', 999)}
""".strip(),
        ))
        
        # Test 2: Dependent code still works (if any)
        dependent_methods = context.get("dependent_methods", [])
        for method in dependent_methods:
            tests.append(TestCase(
                name=f"dependent_{method}_still_works",
                code=f"""
@test isdefined({self.module_name}, :{method})
""".strip(),
            ))
        
        return tests


class AddFieldStrategy(TestGenerationStrategy):
    """Generate tests for ADD_FIELD actions."""
    
    def __init__(self, struct_name: str, field_name: str, field_type: str):
        self.struct_name = struct_name
        self.field_name = field_name
        self.field_type = field_type
    
    def generate(self, context: dict) -> list[TestCase]:
        tests = []
        
        # Test 1: Field exists
        tests.append(TestCase(
            name=f"{self.struct_name}_has_{self.field_name}",
            code=f"""
@test hasfield({self.struct_name}, :{self.field_name})
""".strip(),
        ))
        
        # Test 2: Field type is correct
        tests.append(TestCase(
            name=f"{self.struct_name}_{self.field_name}_type",
            code=f"""
@test fieldtype({self.struct_name}, :{self.field_name}) <: {self.field_type}
""".strip(),
        ))
        
        # Test 3: Struct still constructible
        constructor_args = context.get("constructor_args", "")
        if constructor_args:
            tests.append(TestCase(
                name=f"{self.struct_name}_constructible",
                code=f"""
obj = {self.struct_name}({constructor_args})
@test obj !== nothing
""".strip(),
            ))
        
        return tests


class RenameSymbolStrategy(TestGenerationStrategy):
    """Generate tests for RENAME_SYMBOL actions."""
    
    def __init__(
        self,
        old_name: str,
        new_name: str,
        symbol_type: str = "function",
    ):
        self.old_name = old_name
        self.new_name = new_name
        self.symbol_type = symbol_type
    
    def generate(self, context: dict) -> list[TestCase]:
        tests = []
        module = context.get("module_name", "Main")
        
        # Test 1: New name exists
        tests.append(TestCase(
            name=f"{self.new_name}_exists",
            code=f"""
@test isdefined({module}, :{self.new_name})
""".strip(),
        ))
        
        # Test 2: Old name is gone (unless aliased)
        if not context.get("keep_alias", False):
            tests.append(TestCase(
                name=f"{self.old_name}_gone",
                code=f"""
@test !isdefined({module}, :{self.old_name})
""".strip(),
            ))
        
        return tests


class AddImportStrategy(TestGenerationStrategy):
    """Generate tests for ADD_IMPORT actions."""
    
    def __init__(self, module_name: str, imported_symbols: list[str]):
        self.module_name = module_name
        self.imported_symbols = imported_symbols
    
    def generate(self, context: dict) -> list[TestCase]:
        tests = []
        target_module = context.get("target_module", "Main")
        
        for sym in self.imported_symbols:
            tests.append(TestCase(
                name=f"import_{sym}_available",
                code=f"""
@test isdefined({target_module}, :{sym})
""".strip(),
            ))
        
        return tests


class AddTestStrategy(TestGenerationStrategy):
    """Generate meta-tests for ADD_TEST actions."""
    
    def __init__(self, test_name: str, test_file: str):
        self.test_name = test_name
        self.test_file = test_file
    
    def generate(self, context: dict) -> list[TestCase]:
        tests = []
        
        # Test: The test itself passes
        tests.append(TestCase(
            name=f"{self.test_name}_passes",
            code=f"""
# Run the newly added test
include("{self.test_file}")
@test true  # If we got here, the test passed
""".strip(),
            timeout_ms=30000,
        ))
        
        return tests


# ---------------------------------------------------------------------------
# Main Generator
# ---------------------------------------------------------------------------


class AdaptiveTestGenerator:
    """
    Generate targeted tests based on action type.
    
    This provides richer training signal than running existing tests alone
    by testing the specific consequences of each edit.
    """
    
    STRATEGY_MAP = {
        "ADD_METHOD": AddMethodStrategy,
        "MODIFY_METHOD": ModifyMethodStrategy,
        "REMOVE_METHOD": RemoveMethodStrategy,
        "ADD_FIELD": AddFieldStrategy,
        "MODIFY_FIELD": AddFieldStrategy,  # Similar to add
        "RENAME_SYMBOL": RenameSymbolStrategy,
        "ADD_IMPORT": AddImportStrategy,
        "ADD_TEST": AddTestStrategy,
    }
    
    def generate_tests(
        self,
        action_type: str,
        target: str,
        context: dict,
    ) -> TestSuite:
        """
        Generate a test suite for the given action.
        
        Args:
            action_type: One of ADD_METHOD, MODIFY_METHOD, etc.
            target: Target of the action (method name, field name, etc.)
            context: Additional context needed for test generation
            
        Returns:
            TestSuite with targeted tests for this action.
        """
        strategy_class = self.STRATEGY_MAP.get(action_type)
        
        if strategy_class is None:
            # Return empty suite for unhandled actions
            return TestSuite(action_type=action_type, target=target)
        
        # Create strategy based on action type
        strategy = self._create_strategy(strategy_class, action_type, target, context)
        
        # Generate tests
        tests = strategy.generate(context)
        
        return TestSuite(
            action_type=action_type,
            target=target,
            tests=tests,
            setup_code=context.get("setup_code", ""),
            teardown_code=context.get("teardown_code", ""),
        )
    
    def _create_strategy(
        self,
        strategy_class: type,
        action_type: str,
        target: str,
        context: dict,
    ) -> TestGenerationStrategy:
        """Create strategy instance based on action type."""
        if action_type == "ADD_METHOD":
            return strategy_class(
                method_name=target,
                module_name=context.get("module_name", "Main"),
                arg_types=context.get("arg_types", []),
                return_type=context.get("return_type"),
            )
        elif action_type == "MODIFY_METHOD":
            return strategy_class(
                method_name=target,
                module_name=context.get("module_name", "Main"),
                original_signature=context.get("original_signature", ""),
                new_signature=context.get("new_signature", ""),
            )
        elif action_type == "REMOVE_METHOD":
            return RemoveMethodStrategy(
                method_name=target,
                module_name=context.get("module_name", "Main"),
            )
        elif action_type in ["ADD_FIELD", "MODIFY_FIELD"]:
            return strategy_class(
                struct_name=context.get("struct_name", ""),
                field_name=target,
                field_type=context.get("field_type", "Any"),
            )
        elif action_type == "RENAME_SYMBOL":
            return strategy_class(
                old_name=context.get("old_name", ""),
                new_name=target,
                symbol_type=context.get("symbol_type", "function"),
            )
        elif action_type == "ADD_IMPORT":
            return strategy_class(
                module_name=context.get("source_module", ""),
                imported_symbols=context.get("imported_symbols", [target]),
            )
        elif action_type == "ADD_TEST":
            return strategy_class(
                test_name=target,
                test_file=context.get("test_file", ""),
            )
        else:
            # Shouldn't reach here due to STRATEGY_MAP check
            raise ValueError(f"Unknown action type: {action_type}")
    
    def generate_property_tests(
        self,
        method_name: str,
        properties: list[str],
        context: dict,
    ) -> list[TestCase]:
        """
        Generate property-based tests.
        
        Properties can include: idempotent, commutative, associative, invertible
        """
        tests = []
        module = context.get("module_name", "Main")
        sample = context.get("sample_input", "1")
        
        for prop in properties:
            if prop == "idempotent":
                tests.append(TestCase(
                    name=f"{method_name}_idempotent",
                    code=f"""
x = {sample}
@test {module}.{method_name}({module}.{method_name}(x)) == {module}.{method_name}(x)
""".strip(),
                    timeout_ms=10000,
                ))
            elif prop == "commutative":
                samples = context.get("sample_inputs", "1, 2")
                tests.append(TestCase(
                    name=f"{method_name}_commutative",
                    code=f"""
x, y = {samples}
@test {module}.{method_name}(x, y) == {module}.{method_name}(y, x)
""".strip(),
                    timeout_ms=10000,
                ))
            elif prop == "associative":
                samples = context.get("sample_inputs", "1, 2, 3")
                tests.append(TestCase(
                    name=f"{method_name}_associative",
                    code=f"""
x, y, z = {samples}
@test {module}.{method_name}({module}.{method_name}(x, y), z) == {module}.{method_name}(x, {module}.{method_name}(y, z))
""".strip(),
                    timeout_ms=10000,
                ))
        
        return tests


# ---------------------------------------------------------------------------
# Utility for Julia interop
# ---------------------------------------------------------------------------


def test_suite_to_julia(suite: TestSuite) -> str:
    """Convert a TestSuite to Julia test code."""
    lines = [
        "using Test",
        "",
    ]
    
    if suite.setup_code:
        lines.append("# Setup")
        lines.append(suite.setup_code)
        lines.append("")
    
    lines.append(f'@testset "{suite.action_type}: {suite.target}" begin')
    
    for test in suite.tests:
        lines.append(f'    @testset "{test.name}" begin')
        # Indent the test code
        for line in test.code.split("\n"):
            lines.append(f"        {line}")
        lines.append("    end")
    
    lines.append("end")
    
    if suite.teardown_code:
        lines.append("")
        lines.append("# Teardown")
        lines.append(suite.teardown_code)
    
    return "\n".join(lines)
